cmd/charm-ingest/whitelist.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the GPLv3, see LICENCE file for details.
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/juju/charmstore-client/internal/ingest"
+	"gopkg.in/juju/charmrepo.v4/csclient/params"
+)
+
+var validChannels = params.ValidChannels
+
+// parseWhitelist takes an io.Reader for a whitelist and returns []ingest.WhitelistEntity
+// or an error if unable to parse
+func parseWhitelist(r io.Reader) ([]ingest.WhitelistEntity, error) {
+	var entities []ingest.WhitelistEntity
+	fileScanner := bufio.NewScanner(r)
+	lineNum := 0
+
+	for fileScanner.Scan() {
+		lineNum++
+
+		entity, err := parseLine(fileScanner.Text(), lineNum)
+		if err != nil {
+			return nil, err
+		}
+
+		if entity.EntityId == "" {
+			// Empty line, ignore
+			continue
+		}
+		entities = append(entities, entity)
+	}
+
+	return entities, nil
+}
+
+// parseLine takes a string line of a whitelist file with its corresponding line number,
+// and returns the ingest.WhitelistEntity for that line, or an error
+func parseLine(line string, lineNum int) (ingest.WhitelistEntity, error) {
+	var entity ingest.WhitelistEntity
+
+	fields := strings.Fields(line)
+	if len(fields) == 0 {
+		// Empty line
+		return entity, nil
+	}
+
+	entity.EntityId = fields[0]
+	entity.Channels = make([]params.Channel, len(fields)-1)
+	for i, channelStr := range fields[1:] {
+		channel := params.Channel(channelStr)
+		if !validChannels[channel] {
+			return entity, fmt.Errorf("invalid channel %q for charm %q (line %d)", channelStr, entity.EntityId, lineNum)
```

Comment:
s/charm/entity/
or
s/charm/charm or bundle/
?

(a whitelist line can legitimately refer to a bundle as well as a charm).